querier.go

```diff
@@ -0,0 +1,365 @@
+
+	for _, tr := range q.queryTimeRanges {
+		rangeEnd := time.Now()
+		rangeStart := rangeEnd.Add(-1 * tr)
+
+		queryStart := time.Now()
+		resp, err := q.client.QueryRange(ctx, connect.NewRequest(&queryv1alpha1.QueryRangeRequest{
+			Query: profileType,
+			Start: timestamppb.New(rangeStart),
+			End:   timestamppb.New(rangeEnd),
+		}))
+		latency := time.Since(queryStart)
+		if err != nil {
+			q.metrics.rangeHistogram.WithLabelValues(
+				connect.CodeOf(err).String(), tr.String(),
+			).Observe(latency.Seconds())
+			log.Printf("range(type=%s,over=%s): failed to make request: %v\n", profileType, tr, err)
+			continue
+		}
+
+		q.metrics.rangeHistogram.WithLabelValues(grpcCodeOK, tr.String()).Observe(latency.Seconds())
+		log.Printf(
+			"range(type=%s,over=%s): took %s and got %d series\n",
+			profileType, tr, latency, len(resp.Msg.Series),
+		)
+
+		for k := range q.series {
+			delete(q.series, k)
+		}
+		for _, series := range resp.Msg.Series {
+			if len(series.Samples) < 2 {
+				continue
+			}
+			lSet := labels.Labels{}
+			for _, l := range series.Labelset.Labels {
+				lSet = append(lSet, labels.Label{Name: l.Name, Value: l.Value})
+			}
+			sort.Sort(lSet)
+			q.series[profileType+lSet.String()] = timestampRange{
+				first: series.Samples[0].Timestamp.AsTime(),
+				last:  series.Samples[len(series.Samples)-1].Timestamp.AsTime(),
+			}
+		}
+	}
+}
+
+func (q *Querier) querySingle(ctx context.Context) {
+	if len(q.series) == 0 {
+		log.Println("single: no series to query")
+		return
+	}
+	seriesKeys := maps.Keys(q.series)
+	series := seriesKeys[len(seriesKeys)/2]
+	sampleTs := q.series[series].first
+
+	for _, reportType := range q.reportTypes {
+		queryStart := time.Now()
+		_, err := q.client.Query(ctx, connect.NewRequest(&queryv1alpha1.QueryRequest{
+			Mode: queryv1alpha1.QueryRequest_MODE_SINGLE_UNSPECIFIED,
+			Options: &queryv1alpha1.QueryRequest_Single{
+				Single: &queryv1alpha1.SingleProfile{
+					Query: series,
+					Time:  timestamppb.New(sampleTs),
+				},
+			},
+			ReportType: reportType,
+		}))
+		latency := time.Since(queryStart)
+		if err != nil {
+			q.metrics.singleHistogram.WithLabelValues(
+				connect.CodeOf(err).String(),
+				reportType.String(),
+				"",
+			).Observe(latency.Seconds())
+			log.Printf("single(series=%s, reportType=%s): failed to make request: %v\n", series, reportType, err)
+			continue
+		}
+
+		q.metrics.singleHistogram.WithLabelValues(
+			grpcCodeOK,
+			reportType.String(),
+			"",
+		).Observe(latency.Seconds())
+		log.Printf("single(series=%s, reportType=%s): took %v\n", series, reportType, latency)
+	}
+}
+
+func (q *Querier) queryMerge(ctx context.Context) {
+	if len(q.profileTypes) == 0 {
+		log.Println("merge: no profile types to query")
+		return
+	}
+
+	profileType := q.profileTypes[len(q.profileTypes)/2]
+	for _, tr := range q.queryTimeRanges {
+		if tr > 1*time.Hour {
+			// TODO(asubiotto): This currently OOMs frostdb. Gradually remove
+			// this skip.
+			continue
+		}
```

Comment:
It should, however, be totally fine for Parca as it runs on demo.parca.dev. I would leave this part of parca-load and worst case make it configurable to be turned off, or we deploy a forked version for cloud until cloud isn't OOMing anymore.

Reply:
I'll change it to 24h but I would like to leave this skip in. No parca user will use load generation for queries longer than this time. We should definitely eventually make this configurable (e.g. via https://github.com/parca-dev/parca-load/issues/8), but I think it's a heavier lift than necessary right now. We were previously refusing to issue these queries because we would never have a big enough time range.